pgxn/neon/libpagestore.c

```diff
@@ -427,34 +607,26 @@ pageserver_connect(shardno_t shard_no, int elevel)
 				{
 					char	   *msg = pchomp(PQerrorMessage(conn));
 
-					PQfinish(conn);
-					FreeWaitEventSet(wes);
-
+					CLEANUP_AND_DISCONNECT(shard);
 					neon_shard_log(shard_no, elevel, "could not complete handshake with pageserver: %s",
 								   msg);
-					/* Returning from inside PG_TRY is bad, so we break/return later */
-					broke_from_loop = true;
-					break;
+					pfree(msg);
+					return false;
 				}
 			}
 		}
-	}
-	PG_CATCH();
-	{
-		PQfinish(conn);
-		FreeWaitEventSet(wes);
-		PG_RE_THROW();
-	}
-	PG_END_TRY();
 
-	if (broke_from_loop)
-	{
-		return false;
+		shard->state = PS_Connected;
+		/* fallthrough */
+	}
+	case PS_Connected:
+		neon_shard_log(shard_no, LOG, "libpagestore: connected to '%s' with protocol version %d", connstr, neon_protocol_version);
+		return true;
```

Comment:
Not new with this PR, but shoudn't we call PQgetResult() here? Which should return PGRES_COPY_BOTH.